funcx_endpoint/funcx_endpoint/endpoint/result_store.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import annotations
+
+import os
+import pathlib
+import sys
+import typing as t
+
+
+class ResultStore:
+    """
+    Persists bytes to disk by key via a dict()-like API.
+
+    The ResultStore is a generic byte-storage data structure that stores and
+    retrieves bytes to the filesystem.  Each payload is stored on the
+    filesystem as a single file, named by the associated key.  Payloads are
+    stored in a subdirectory of the init path, named "unacked_results".
+
+    A typical interaction might look like:
+
+        >>> rs = ResultStore(endpoint_dir="some_dir")
+        >>> rs["some_id_01"] = b"some_payload_01"
+
+    An alternate method to store results:
+
+        >>> rs += ("some_id_02", b"some_payload_02")
+
+    Access a single result by the result key
+
+            # will raise if key does not exist
+        >>> storage_time, stored_bytes = rs["some_id_01"]
+
+            # will return a default value of None, if nkey
+        >>> storage_time, stored_bytes = rs.get("some_id_01") # return a default
+
+    Iterate all currently stored items via iteration:
+
+        >>> for task_id, ctime, stored_bytes in rs:
+
+    Discard stored items via `.pop()`, `.remove()`, or `.discard()`
+
+        >>> rs.pop("some_id_01")  # also returns result, but ignored here
+        >>> rs.pop("some_id_01")  # will raise as key does not exist
+        >>> rs.remove("some_id_02")
+        >>> rs.remove("some_id_02")  # will raise as key does not exist
+        >>> rs.discard("some_id_02")  # will not raise if key does not exist
+
+    Completely empty the ResultStore with `.clear()`.  All non-hidden files
+    (those not beginning with a dot) in the unacked_results directory will be
+    removed from the filesystem.
+
+        >>> rs.clear()
+    """
+
+    def __init__(self, endpoint_dir: str | pathlib.Path):
+        self.endpoint_dir = pathlib.Path(endpoint_dir)
+        self.data_path = self.endpoint_dir / "unacked_results"
+        self.data_path.mkdir(exist_ok=True)
+
+        self.data_path.chmod(mode=0o0700)
+        test_path = self.data_path / ".test-reading-and-writing.txt"
+        test_path.write_bytes(b"Verify write-ability")
+        if not any(i for i in self.data_path.glob(".test*")):
+            msg = (
+                f"Unable to list files; does {self.data_path} have the correct"
+                " permissions?"
+            )
+            raise PermissionError(msg)
+        test_path.unlink()
+
+    def __contains__(self, key: str) -> bool:
+        normpath = os.path.normpath(key)
+        return (self.data_path / normpath).exists()
+
+    def __iter__(self) -> t.Iterable[tuple[str, float, bytes]]:
+        for rp in self._iter_result_paths():
+            yield rp.name, rp.lstat().st_ctime, rp.read_bytes()
```

Comment:
Not sure what benefit we get from return ctime in the iterator. Are we using this maybe in tests?

Reply:
This is a holdover from my initial thought-process of "what do we do if we have a result that _just won't send_?"  I was unable to come up with a specific error case that enabled this, but was trying to avoid any possibility that stale results could build up -- the logic that was initially in place checked that if a result was more than STALE_RESULT_TTL (a config option, defaulting to 2 weeks) old, it was dropped.

I'm having a difficult time fully letting go of the original thought-process because I don't entirely agree that this will be non-issue.  However, I suppose we could always add it in again later if we change our mind.

Reply:
I understand. It's simpler to leave out bits that are not in use right away. I'll leave it up to you.

Reply:
I removed it.